rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,392 @@
+
+If a Teleport Client notices that a slot is in use, but not by a Teleport client, then it should prompt the user before overwriting the slot. This will also be done by checking the certificate stored in the slot. We will share the certificate in the prompt, since it may contain identifying information. e.g:
+
+```
+> tsh login
+Hardware key slot 9a appears to be in use by another program.
+Slot 9a:
+	Algorithm:	ECCP256
+	Subject DN:	CN=SSH key
+	Issuer DN:	OU=(devel),O=yubikey-agent
+	Serial:		20876611871300106558747702921785395021
+	Fingerprint:	1ce4faf8bdbfc9668a9f532c20b03ccf1dbadcd06b51f235aeb3fe388bb1703b
+	Not before:	2022-08-19 01:10:14
+	Not after:	2064-08-19 01:10:14
+Would you like to overwrite this slot? (y/N):
+```
+
+#### Private key interface
+
+Currently, Teleport clients store a PEM encoded private key (`~/.tsh/keys/proxy/user`) for a login session. This PEM encoded private key is then unmarshalled, transformed, and parsed as needed during a client request.
+
+With a hardware private key, we only have access to a raw `crypto.PrivateKey`, and do not have sufficient information about the key to transform it into an `*rsa.PrivateKey` and marshal it into PKCS1 format. Instead, we need to alter Teleport clients to use `cyrpto.PrivateKey` by default. This will require altering the key interface (`lib/client/interfaces.go`) and its usage across `lib/client` and other relevant locations. `lib/utils/native` will also be updated to return `*rsa.PrivateKey` instead of its PEM encoded private and public keys.
+
+We also need a way for future Teleport Client requests to retrieve the correct `crypto.PrivateKey`. For RSA keys, we can continue to store them as PEM encoded keys in (`~/.tsh/keys/proxy/user`). For hardware private keys, we will instead store a fake PEM encoded private key which we can use to identity what device and slot to load the private key from.
+
+```
+-----BEGIN YUBIKEY PIV PRIVATE KEY-----
+# PEM encoded
+serial_number=<serial_number>
+slot=<slot>
+-----END YUBIKEY PIV PRIVATE KEY-----
+```
+
+#### Supported clients
+
+`tsh` and Teleport Connect will both support hardware private key login, and `tctl` will be able to use resulting login sessions. 
+
+#### Unsupported clients
+
+The WebUI will not be able to support PIV login, since it is brower-based and cannot connect directly to the user's PIV device. If a user with `require_session_mfa: hardware_key` attempts to login on the WebUI, or use an existing login session, it will fail. However, WebUI user registration and password reset logic must still work, regardless of the user's private key policy requirement. After intial registration/reset flow, the user should be directed to a page which notifies them that `tsh` or Teleport Connect must be used.
```

Comment:
```suggestion
The WebUI will not be able to support PIV login, since it is brower-based and cannot connect directly to the user's PIV device. If a user with `require_session_mfa: hardware_key` attempts to login on the WebUI, or use an existing login session, it will fail. However, WebUI user registration and password reset logic must still work, regardless of the user's private key policy requirement. After initial registration/reset flow, the user should be directed to a page which notifies them that `tsh` or Teleport Connect must be used.
```